File: testdata/fixtures/kotlin/CWE-502b.kt
// CWE-502: Deserialization via ObjectInputStream
package fixture
import java.io.ObjectInputStream
import java.io.ByteArrayInputStream
import javax.servlet.http.HttpServletRequest

class DeserTest {
    fun vuln(request: HttpServletRequest) {
        val data = request.getParameter("data")
        // ruleid: kotlin-deserialization
        ObjectInputStream(ByteArrayInputStream(data.toByteArray())).readObject()
    }

    fun vulnHeader(request: HttpServletRequest) {
        // ruleid: kotlin-deserialization
        ObjectInputStream(ByteArrayInputStream(request.getHeader("X-Data").toByteArray())).readUnshared()
    }

    fun vulnIntent(intent: android.content.Intent) {
        val extra = intent.getStringExtra("extra")
        // ruleid: kotlin-deserialization
        ObjectInputStream(ByteArrayInputStream(extra.toByteArray())).readObject()
    }

    fun safe() {
        // ok: kotlin-deserialization
        ObjectInputStream(ByteArrayInputStream("fixed".toByteArray())).readObject()
    }
}

File: testdata/fixtures/ocaml/CWE-502b.ml
(* CWE-502: Deserialization *)
let vuln str ch =
  (* ruleid: ocaml-deserialization *)
  let _ = Marshal.from_string str 0 in
  (* ruleid: ocaml-deserialization *)
  let _ = Marshal.from_bytes (Bytes.of_string str) 0 in
  (* ruleid: ocaml-deserialization *)
  let _ = input_value ch in
  ()
let safe () =
  (* ok: ocaml-deserialization *)
  let _ = Marshal.to_string "safe" [] in
  (* ok: ocaml-deserialization *)
  let _ = Marshal.to_bytes "safe" [] in
  ()
